// stream-chat-react/dist/es/audioProcessing.21cb49e1.mjs
import { useCallback } from "react";
const useHandleFileChangeWrapper = (resetOnChange = false, handler) => useCallback(
  ({ currentTarget }) => {
    const { files } = currentTarget;
    if (!files) return;
    try {
      handler?.(Array.from(files));
    } catch (error) {
      console.error(error);
    }
    if (resetOnChange) currentTarget.value = "";
  },
  [handler, resetOnChange]
);
async function dataTransferItemsToFiles(items) {
  if (!items || !items.length) {
    return [];
  }
  const fileLikes = getFileLikes(items);
  if (fileLikes.length) {
    return fileLikes;
  }
  const blobPromises = [];
  for (const item of items) {
    if (item.type === "text/html") {
      blobPromises.push(
        new Promise((accept) => {
          item.getAsString(async (s) => {
            const imagePromises = extractImageSources(s).map(
              (src) => getImageSource(fileLikes, src)
            );
            await Promise.all(imagePromises);
            accept();
          });
        })
      );
    }
  }
  await Promise.all(blobPromises);
  return fileLikes;
}
function getFileLikes(items) {
  const fileLikes = [];
  for (const item of items) {
    if (item.kind === "file") {
      const file = item.getAsFile();
      if (file) {
        fileLikes.push(file);
      }
    }
  }
  return fileLikes;
}
async function getImageSource(fileLikes, src) {
  let res;
  try {
    res = await fetch(src);
  } catch (e) {
    return;
  }
  const contentType = res.headers.get("Content-type") || "application/octet-stream";
  const buf = await res.arrayBuffer();
  const blob = new Blob([buf], { type: contentType });
  fileLikes.push(blob);
}
const extractImageSources = (s) => {
  const imageTags = new DOMParser().parseFromString(s, "text/html").getElementsByTagName("img");
  return Array.from(imageTags, (tag) => tag.src).filter((tag) => tag);
};
const createFileFromBlobs = ({
  blobsArray,
  fileName,
  mimeType
}) => {
  const concatenatedBlob = new Blob(blobsArray, { type: mimeType });
  return new File([concatenatedBlob], fileName, { type: concatenatedBlob.type });
};
const getExtensionFromMimeType = (mimeType) => {
  const match = mimeType.match(/\/([^/;]+)/);
  return match && match[1];
};
const getRecordedMediaTypeFromMimeType = (mimeType) => {
  const match = mimeType.match(/^(audio|video)\/.*$/);
  return match && match[1];
};
const readFileAsArrayBuffer = (file) => new Promise((resolve, reject) => {
  const fileReader = new FileReader();
  fileReader.onload = () => {
    resolve(fileReader.result);
  };
  fileReader.onerror = () => {
    reject(fileReader.error);
  };
  fileReader.readAsArrayBuffer(file);
});
const toAudioBuffer = async (file) => {
  const audioCtx = new AudioContext();
  const arrayBuffer = await readFileAsArrayBuffer(file);
  const decodedData = await audioCtx.decodeAudioData(arrayBuffer);
  if (audioCtx.state !== "closed") await audioCtx.close();
  return decodedData;
};
const renderAudio = async (audioBuffer, sampleRate) => {
  const offlineAudioCtx = new OfflineAudioContext(
    audioBuffer.numberOfChannels,
    audioBuffer.duration * sampleRate,
    sampleRate
  );
  const source = offlineAudioCtx.createBufferSource();
  source.buffer = audioBuffer;
  source.connect(offlineAudioCtx.destination);
  source.start();
  return await offlineAudioCtx.startRendering();
};
export {
  getRecordedMediaTypeFromMimeType as a,
  createFileFromBlobs as c,
  dataTransferItemsToFiles as d,
  getExtensionFromMimeType as g,
  renderAudio as r,
  toAudioBuffer as t,
  useHandleFileChangeWrapper as u
};
